asciidoctorj-core/src/main/java/org/asciidoctor/ast/Document.java

```diff
@@ -1,53 +1,51 @@
 package org.asciidoctor.ast;
 
-import java.util.HashMap;
 import java.util.Map;
 
 import org.asciidoctor.internal.RubyHashUtil;
-import org.asciidoctor.internal.RubyUtils;
-import org.jruby.Ruby;
 import org.jruby.RubyHash;
+import org.jruby.RubyString;
 import org.jruby.runtime.builtin.IRubyObject;
 
 public class Document extends AbstractBlockImpl implements DocumentRuby {
 
-    private DocumentRuby documentDelegate;
-
-    public Document(DocumentRuby documentRuby, Ruby runtime) {
-        super(documentRuby, runtime);
-        this.documentDelegate = documentRuby;
+    public Document(IRubyObject documentRuby) {
+        super(documentRuby);
     }
 
-    public DocumentRuby getDocumentRuby() {
-        return documentDelegate;
+    @Override
+    public boolean isBasebackend(String backend) {
+        return getBoolean("basebackend?", backend);
     }
 
     @Override
     public boolean basebackend(String backend) {
-        return documentDelegate.basebackend(backend);
+        return isBasebackend(backend);
     }
 
     @Override
     public Map<Object, Object> getOptions() {
-        Map<Object, Object> options = (Map<Object, Object>)documentDelegate.getOptions();
-        return RubyHashUtil.convertRubyHashMapToMap(options);
+        return RubyHashUtil.convertRubyHashMapToMap((RubyHash) getRubyProperty("options"));
     }
 
     @Override
     public Object doctitle(Map<Object, Object> opts) {
         RubyHash mapWithSymbols = RubyHashUtil.convertMapToRubyHashWithSymbolsIfNecessary(runtime, opts);
 
-        Object doctitle = documentDelegate.doctitle(mapWithSymbols);
+        Object doctitle = getRubyProperty("doctitle", mapWithSymbols);
 
-        if (doctitle instanceof IRubyObject) {
-            doctitle = RubyUtils.rubyToJava(runtime, (IRubyObject) doctitle, Title.class);
+        if (doctitle instanceof RubyString) {
+            return ((RubyString) doctitle).asJavaString();
+        } else if (doctitle instanceof String) {
+            return doctitle;
+        } else {
+            return toJava((IRubyObject) doctitle, Title.class);
         }
 
-        return doctitle;
     }
 
     public String doctitle() {
-        return (String) doctitle(new HashMap<Object, Object>());
+        return getString("doctitle");
```

Comment:
I'm okay with that. I combined it as a single method in the Ruby side because that's more the Ruby way. But we could totally have a different name in the Java side. Something like getStructuredDoctitle() or something.